hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeAttributeProvider.java

```diff
@@ -17,19 +17,227 @@
+      }
+
+      public Builder ignoreEmptyDir(boolean val) {
+        this.ignoreEmptyDir = val;
+        return this;
+      }
+
+      public Builder operationName(String val) {
+        this.operationName = val;
+        return this;
+      }
+
+      public Builder callerContext(CallerContext val) {
+        this.callerContext = val;
+        return this;
+      }
+    }
+
+    public AuthorizationContext(
+        String fsOwner,
+        String supergroup,
+        UserGroupInformation callerUgi,
+        INodeAttributes[] inodeAttrs,
+        INode[] inodes,
+        byte[][] pathByNameArr,
+        int snapshotId,
+        String path,
+        int ancestorIndex,
+        boolean doCheckOwner,
+        FsAction ancestorAccess,
+        FsAction parentAccess,
+        FsAction access,
+        FsAction subAccess,
+        boolean ignoreEmptyDir) {
+      this.fsOwner = fsOwner;
+      this.supergroup = supergroup;
+      this.callerUgi = callerUgi;
+      this.inodeAttrs = inodeAttrs;
+      this.inodes = inodes;
+      this.pathByNameArr = pathByNameArr;
+      this.snapshotId = snapshotId;
+      this.path = path;
+      this.ancestorIndex = ancestorIndex;
+      this.doCheckOwner = doCheckOwner;
+      this.ancestorAccess = ancestorAccess;
+      this.parentAccess = parentAccess;
+      this.access = access;
+      this.subAccess = subAccess;
+      this.ignoreEmptyDir = ignoreEmptyDir;
+    }
+
+    public AuthorizationContext(
+        String fsOwner,
+        String supergroup,
+        UserGroupInformation callerUgi,
+        INodeAttributes[] inodeAttrs,
+        INode[] inodes,
+        byte[][] pathByNameArr,
+        int snapshotId,
+        String path,
+        int ancestorIndex,
+        boolean doCheckOwner,
+        FsAction ancestorAccess,
+        FsAction parentAccess,
+        FsAction access,
+        FsAction subAccess,
+        boolean ignoreEmptyDir,
+        String operationName,
+        CallerContext callerContext) {
+      this(fsOwner, supergroup, callerUgi, inodeAttrs, inodes,
+          pathByNameArr, snapshotId, path, ancestorIndex, doCheckOwner,
+          ancestorAccess, parentAccess, access, subAccess, ignoreEmptyDir);
+      this.operationName = operationName;
+      this.callerContext = callerContext;
+    }
+
+    public AuthorizationContext(Builder builder) {
+      this(builder.fsOwner, builder.supergroup, builder.callerUgi,
+          builder.inodeAttrs, builder.inodes, builder.pathByNameArr,
+          builder.snapshotId, builder.path, builder.ancestorIndex,
+          builder.doCheckOwner, builder.ancestorAccess, builder.parentAccess,
+          builder.access, builder.subAccess, builder.ignoreEmptyDir);
+      this.operationName = builder.operationName;
+      this.callerContext = builder.callerContext;
+    }
+
+    @VisibleForTesting
+    @Override
+    public boolean equals(Object obj) {
+      if (!(obj instanceof AuthorizationContext)) {
+        return false;
+      }
+      return true;
```

Comment:
I don't think it is right to return always return true for two different AuthorizationContext instances.

Reply:
yes... it was meant only for test. Updated the patch to include the full object equivalence check.